streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java

```diff
@@ -105,14 +100,28 @@ public boolean persistent() {
     }
 
     @Override
-    public V get(K key) {
-        return this.map.get(key);
+    public boolean open() {
+        return open;
+    }
+
+    @Override
+    public  V get(K key) {
+        lock.readLock().lock();
+        try {
+            return this.map.get(key);
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
     public void put(K key, V value) {
-        this.map.put(key, value);
-        this.keys.add(key);
+        lock.writeLock().lock();
```

Comment:
I just read through `MemoryLRUCache`. It is not thread-safe and will corrupt itself because a read causes a mutation of the LRU history. (I made the same mistake early in my career when fixing performance problems leading to exploring caching in-depth, so its an easy oversight to make)

A read/write lock is a very expensive mechanism and most often the incorrect lock type to use. For short critical sections it is more expensive than an exclusive lock. By using a `ReentrantLock` or `synchronized` you'll have both correctness and higher performance.

As is, I strongly urge you to correct this before merging. You don't have to use a caching library, but the code is very broken.